cpp/src/groupby/hash/compute_single_pass_aggs.cu

```diff
@@ -0,0 +1,132 @@
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+namespace cudf::groupby::detail::hash {
+// make table that will hold sparse results
+auto create_sparse_results_table(table_view const& flattened_values,
+                                 std::vector<aggregation::Kind> aggs,
+                                 rmm::cuda_stream_view stream)
+{
+  // TODO single allocation - room for performance improvement
+  std::vector<std::unique_ptr<column>> sparse_columns;
+  sparse_columns.reserve(flattened_values.num_columns());
+  std::transform(
+    flattened_values.begin(),
+    flattened_values.end(),
+    aggs.begin(),
+    std::back_inserter(sparse_columns),
+    [stream](auto const& col, auto const& agg) {
+      bool nullable =
+        (agg == aggregation::COUNT_VALID or agg == aggregation::COUNT_ALL)
+          ? false
+          : (col.has_nulls() or agg == aggregation::VARIANCE or agg == aggregation::STD);
+      auto mask_flag = (nullable) ? mask_state::ALL_NULL : mask_state::UNALLOCATED;
+
+      auto col_type = cudf::is_dictionary(col.type())
+                        ? cudf::dictionary_column_view(col).keys().type()
+                        : col.type();
+
+      return make_fixed_width_column(
+        cudf::detail::target_type(col_type, agg), col.size(), mask_flag, stream);
+    });
+
+  table sparse_table(std::move(sparse_columns));
+  mutable_table_view table_view = sparse_table.mutable_view();
+  cudf::detail::initialize_with_identity(table_view, aggs, stream);
+  return sparse_table;
+}
+
+/**
+ * @brief Computes all aggregations from `requests` that require a single pass
+ * over the data and stores the results in `sparse_results`
+ */
+template <typename SetType>
+void compute_single_pass_aggs(int64_t num_keys,
+                              bool skip_rows_with_nulls,
+                              bitmask_type const* row_bitmask,
+                              SetType set,
+                              host_span<aggregation_request const> requests,
+                              cudf::detail::result_cache* sparse_results,
+                              rmm::cuda_stream_view stream)
+{
+  // flatten the aggs to a table that can be operated on by aggregate_row
+  auto const [flattened_values, agg_kinds, aggs] = flatten_single_pass_aggs(requests);
+
+  // make table that will hold sparse results
+  table sparse_table = create_sparse_results_table(flattened_values, agg_kinds, stream);
+  // prepare to launch kernel to do the actual aggregation
+  auto d_sparse_table = mutable_table_device_view::create(sparse_table, stream);
+  auto d_values       = table_device_view::create(flattened_values, stream);
+  auto const d_aggs   = cudf::detail::make_device_uvector_async(
+    agg_kinds, stream, cudf::get_current_device_resource_ref());
+
+  thrust::for_each_n(
+    rmm::exec_policy_nosync(stream),
+    thrust::make_counting_iterator(0),
```

Comment:
Should either add a `CUDF_EXPECTS(num_keys <= std::numeric_limits<size_type>::max(), "...");` before this `for_each_n` or making counting_iterator int64 type as `num_keys` can be > 2B.

```suggestion
    thrust::make_counting_iterator<int64_t>(0),
```

Would prefer the former one as the `compute_single_pass_aggs_fn` functor works with `size_type` indexes.

Reply:
Nice you caught this. This is an interesting situation:

- Theoretically, `num_keys` will never exceed 2 billion since it is determined by the number of rows in a table and is therefore of type `cudf::size_type`.
- We use `int64_t` for the hash set extent (size) type because libcudf, in this case, uses a cuco hash table with 50% occupancy. For instance, with an input of 1.5 billion rows, the required capacity would be 1.5B x 2 = 3B, which exceeds the limit of `int32_t`.

Once the hash set is created, we need to insert and query using row indices, so using the default `size_type` is indeed the correct approach.

Reply:
Makes sense. I guess we can skip the redundant `CUDF_EXPECTS` too in this case as it is bound by the number of rows.